### ISSUE TITLE
Syntax for aggregation of references in bibliography

### ISSUE THREAD

**Issue by spl** [jgm/pandoc-citeproc#142]
I would like to generate, in HTML output, an aggregation of citations from a BibTeX file. For example, given a query for all citations that match an author's name or have a certain keyword or fall within a range of publication dates, those citations will be rendered in a generated bibliography. (Ideally, I think, the query would be determined by some syntax in a source Markdown file.)

Is there some pre-existing way to do this? Or are there any suggestions for how to implement it?

**Comment by jgm** [comment]
You can generate a bibliography that contains everything
in a BibTeX file by doing this in the YAML metadata:

   ---
   nocite: '@*'
   ...

There's no way to pattern-match citations, though; you can
specify them individually or use the wildcard.  There are
undoubtedly bibtex tools you could use to separate out all
the bibtex entries matching a condition, and create a new
bibtex file.

+++ Sean Leather [Jun 19 15 06:15 ]:

**Comment by spl** [comment]
(Just FYI, your email response isn't rendering nicely in GitHub, so I'm including the relevant parts in my response below.)

> You can generate a bibliography that contains everything in a BibTeX file by doing this in the YAML metadata:
>
> ```
> ---
>    nocite: '@*'
> ...
> ```

Okay, that's useful.

> There are undoubtedly bibtex tools you could use to separate out all the bibtex entries matching a condition, and create a new bibtex file.

Indeed. I found one [list](http://tex.stackexchange.com/questions/417/how-to-split-all-bibtex-referenced-entries-from-a-big-bibtex-database-to-a-copy), including [BibTool](http://www.ctan.org/tex-archive/biblio/bibtex/utils/bibtool/).

> There's no way to pattern-match citations, though; you can specify them individually or use the wildcard.

Supposing, however, that I prefer to keep it all “in-house” (for a website generated using Hakyll, a `bibliography.bib`, and a `style.csl`), what do you think of the following idea?

Based on some YAML metadata including a filter in a Markdown file, I read the file and filter the `refs :: [Reference]` obtained from `bibliography.bib`. I then pass the filtered `refs` as the `references` YAML metadata to [`Hakyll.Web.Pandoc.writePandocWith`](http://hackage.haskell.org/package/hakyll-4.7.0.0/docs/Hakyll-Web-Pandoc.html#v:writePandocWith), which uses [`Text.Pandoc.Writers.HTML.writeHtmlString`](http://hackage.haskell.org/package/pandoc-1.14.0.4/docs/Text-Pandoc-Writers-HTML.html#v:writeHtmlString).

I'm currently working on an approach like that. Does it make sense?

**Comment by jgm** [comment]
+++ Sean Leather [Jun 23 15 00:01 ]:
>   Based on some YAML metadata including a filter in a Markdown file, I
>   read the file and filter the refs :: [Reference] obtained from
>   bibliography.bib. I then pass the filtered refs as the references YAML
>   metadata to [3]Hakyll.Web.Pandoc.writePandocWith, which uses
>   [4]Text.Pandoc.Writers.HTML.writeHtmlString.
>
>   I'm currently working on an approach like that. Does it make sense?

Sounds reasonable to me!

**Comment by spl** [comment]
I've learned that Hakyll does not use Pandoc's metadata. It reads the YAML metadata itself and uses it separately from Pandoc. So it doesn't work to write out the references as Pandoc metadata. Rather, I use Hakyll's template context.

I do use filters and sorting on `[Reference]` to do aggregations on citations. It still might be useful to have syntax for doing that in the YAML. Not yet sure what it would look like, though.

**Comment by spl** [comment]
In the project I was working on, we ended up needing more than a query on references. We also needed to modify the rendered HTML output (by adding stuff to the rendering of each reference in the bibliography). `Text.CSL.Pandoc.processCites` does not directly allow this, so we instead hacked it by rendering dummy, one-`nocite`-citation-only bibliography HTML strings and combining these strings with other strings to make our bibliographies. It would be nice if there was a way to do more manipulation of the bibliography in pandoc-citeproc, but I can certainly understand that this is not a primary use case.

I'll close this issue for now since I don't expect to need it or work on it.
